// electron-builder-lib/out/targets/differentialUpdateInfoBuilder.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.createBlockmap = exports.appendBlockmap = exports.BLOCK_MAP_FILE_SUFFIX = undefined;

var _bluebirdLst;

function _load_bluebirdLst() {
    return _bluebirdLst = require("bluebird-lst");
}

let appendBlockmap = exports.appendBlockmap = (() => {
    var _ref = (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* (file) {
        return JSON.parse((yield (0, (_util || _load_util()).exec)((yield getBlockMapTool()), ["-in", file, "-append", "-compression", "deflate"])));
    });

    return function appendBlockmap(_x) {
        return _ref.apply(this, arguments);
    };
})();

let createBlockmap = exports.createBlockmap = (() => {
    var _ref2 = (0, (_bluebirdLst || _load_bluebirdLst()).coroutine)(function* (file, target, packager, safeArtifactName) {
        const blockMapFile = `${file}${BLOCK_MAP_FILE_SUFFIX}`;
        const updateInfo = JSON.parse((yield (0, (_util || _load_util()).exec)((yield getBlockMapTool()), ["-in", file, "-out", blockMapFile])));
        packager.info.dispatchArtifactCreated({
            file: blockMapFile,
            safeArtifactName: `${safeArtifactName}${BLOCK_MAP_FILE_SUFFIX}`,
            target,
            arch: null,
            packager,
            updateInfo
        });
        return updateInfo;
    });

    return function createBlockmap(_x2, _x3, _x4, _x5) {
        return _ref2.apply(this, arguments);
    };
})();


exports.createNsisWebDifferentialUpdateInfo = createNsisWebDifferentialUpdateInfo;
exports.configureDifferentialAwareArchiveOptions = configureDifferentialAwareArchiveOptions;

var _util;

function _load_util() {
    return _util = require("builder-util/out/util");
}

var _path = _interopRequireWildcard(require("path"));

var _tools;

function _load_tools() {
    return _tools = require("./tools");
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

const BLOCK_MAP_FILE_SUFFIX = exports.BLOCK_MAP_FILE_SUFFIX = ".blockmap";
function createNsisWebDifferentialUpdateInfo(artifactPath, packageFiles) {
    if (packageFiles == null) {
        return null;
    }
    const keys = Object.keys(packageFiles);
    if (keys.length <= 0) {
        return null;
    }
    const packages = {};
    for (const arch of keys) {
        const packageFileInfo = packageFiles[arch];
        packages[arch] = Object.assign({}, packageFileInfo, { path: _path.basename(packageFileInfo.path) });
    }
    return { packages };
}
function configureDifferentialAwareArchiveOptions(archiveOptions) {
    archiveOptions.solid = false;
    // our reader doesn't support compressed headers
    archiveOptions.isArchiveHeaderCompressed = false;
    /*
     * dict size 64 MB: Full: 33,744.88 KB, To download: 17,630.3 KB (52%)
     * dict size 16 MB: Full: 33,936.84 KB, To download: 16,175.9 KB (48%)
     * dict size  8 MB: Full: 34,187.59 KB, To download:  8,229.9 KB (24%)
     * dict size  4 MB: Full: 34,628.73 KB, To download: 3,782.97 KB (11%)
        as we can see, if file changed in one place, all block is invalidated (and update size approximately equals to dict size)
     */
    archiveOptions.dictSize = 4;
    // do not allow to change compression level to avoid different packages
    archiveOptions.compression = "normal";
    return archiveOptions;
}
function getBlockMapTool() {
    // noinspection SpellCheckingInspection
    return (0, (_tools || _load_tools()).getTool)({
        repository: "develar/block-map-builder",
        name: "block-map-builder",
        version: "0.1.0",
        mac: "hqlM948NJFglvJ9S0OC4COcc1rw2CwYjc1KJQKED/PfKHhPQsLbv5Z1Mi13mP5C/g/6bXqTnwjs2gLYTocfTVQ==",
        "linux-ia32": "E7YpTsavXC9vZ0NZV6pykpvgYhHxyQBJsoLCHIWwwQA9mrSpnq828qCdE5c4/aMsoETlwxBWe6BFObSn4dzSPw==",
        "linux-x64": "08Ps5UWzeUrfOBDwu3QGjH/QcHeFHnfURdrl8OpXNGLfYFMgobFGbIMw3YY+Jc+YFofpvruRoWDJ0yBGXWHr9w==",
        "win-ia32": "63xaYaZl/8LAwGigfVBgeDWEhZeViOvv4tRetkeppD4qCohwCRgoMI1l9+k5LeyqOnjethFzkJASZfbGi6Ww+Q==",
        "win-x64": "ABcfppB8I5O2A1L1ZwpRjCXqHFMuqoeeNCE+mhR9i9nKJANz2peS0Ob6sD7Y5M2R0TYDNLyQCC4YCexk6NWVHw=="
    });
}
